File: app/src/main/java/com/legue/axel/lolsummonertool/adapter/ChampionSpellAdapter.java
package com.legue.axel.lolsummonertool.adapter;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.text.Html;
import android.text.TextUtils;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.DataSource;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.engine.GlideException;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.legue.axel.lolsummonertool.R;
import com.legue.axel.lolsummonertool.database.model.champion.Spell;
import com.legue.axel.lolsummonertool.database.model.champion.SpellImage;
import com.legue.axel.lolsummonertool.database.viewmodel.SpellViewModel;
import com.legue.axel.lolsummonertool.utils.ImageUtils;
import com.legue.axel.lolsummonertool.utils.Utils;
import com.legue.axel.lolsummonertool.wiki.activity.WikiChampionInformations;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ChampionSpellAdapter extends RecyclerView.Adapter<ChampionSpellAdapter.ChampionSpellHolder> {
    private static final String TAG = ChampionSpellAdapter.class.getName();

    private Context mContext;
    private List<Spell> mSpells;
    private WikiChampionInformations mActivity;
    private SpellImage mSpellImage;

    public ChampionSpellAdapter(Context mContext, List<Spell> mSpells, WikiChampionInformations mActivity) {
        this.mContext = mContext;
        this.mSpells = mSpells;
        this.mActivity = mActivity;
    }

    @NonNull
    @Override
    public ChampionSpellHolder onCreateViewHolder(@NonNull ViewGroup parent, int i) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.adapter_champion_spell, parent, false);
        return new ChampionSpellAdapter.ChampionSpellHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull ChampionSpellHolder holder, int i) {
        final Spell spell = mSpells.get(i);

        if (spell != null) {

            SpellViewModel spellViewModel = ViewModelProviders.of(mActivity).get(SpellViewModel.class);
            spellViewModel.getSpellImage(spell.id).observe(mActivity, spellImage -> {
                if (spellImage != null) {
                    mSpellImage = spellImage;
                    displayImage(mSpellImage.full, holder.ivSpell, holder.pbSpell);
                }
            });

            //TODO complete info


            if (spell.name != null && !TextUtils.isEmpty(spell.name)) {
                holder.tvName.setText(spell.name);

                if (shouldConvertListToString(spell.cost)) {
                    holder.tvCost.setText(Utils.converFloatListToString(spell.cost));
                } else {
                    holder.tvCost.setText(String.valueOf(Math.round(spell.cost.get(0))));
                }

                if (shouldConvertListToString(spell.range)) {
                    holder.tvRange.setText(Utils.converFloatListToString(spell.range));
                } else {
                    holder.tvRange.setText(String.valueOf(Math.round(spell.range.get(0))));
                }
                holder.tvCooldown.setText(Utils.converFloatListToString(spell.cooldown));
                holder.tvLore.setText(Html.fromHtml(spell.description, Html.FROM_HTML_MODE_COMPACT));
            }


        } else {
            Log.i(TAG, "onBindViewHolder: spell is null");
        }

    }


    @Override
    public int getItemCount() {
        return mSpells.size();
    }

    private boolean shouldConvertListToString(List<Float> list) {
        Float value = list.get(0);
        for (int i = 0; i < list.size(); i++) {
            if (value.equals(list.get(i))) {
                //do nothing
            } else {
                return true;
            }
        }
        return false;
    }


    private void displayImage(String url, ImageView imageView, ProgressBar progressBar) {
        if (mSpellImage != null) {
            Glide.with(mContext)
                    .load(ImageUtils.BuildSpellIconUrl(url))
                    .listener(new RequestListener<Drawable>() {
                        @Override
                        public boolean onLoadFailed(@Nullable GlideException e, Object model, Target<Drawable> target, boolean isFirstResource) {
                            progressBar.setVisibility(View.GONE);
                            return false;
                        }

                        @Override
                        public boolean onResourceReady(Drawable resource, Object model, Target<Drawable> target, DataSource dataSource, boolean isFirstResource) {
                            progressBar.setVisibility(View.GONE);
                            imageView.setVisibility(View.VISIBLE);
                            return false;
                        }
                    })
                    .error(R.drawable.ic_placeholder_black_24dp)
                    .diskCacheStrategy(DiskCacheStrategy.RESOURCE)
                    .into(imageView);
        } else {
            Log.i(TAG, "displayImage: null");
        }

    }

    class ChampionSpellHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.iv_spell)
        ImageView ivSpell;

        @BindView(R.id.tv_spell_name)
        TextView tvName;
        @BindView(R.id.tv_range)
        TextView tvRange;
        @BindView(R.id.tv_cost)
        TextView tvCost;
        @BindView(R.id.tv_cooldown)
        TextView tvCooldown;
        @BindView(R.id.tv_description)
        TextView tvLore;
        @BindView(R.id.pb_spell)
        ProgressBar pbSpell;

        ChampionSpellHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/com/legue/axel/lolsummonertool/adapter/ChampionSpellAdapter.kt
package com.legue.axel.lolsummonertool.adapter

import android.content.Context
import android.graphics.drawable.Drawable
import android.text.Html
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.bumptech.glide.load.DataSource
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.engine.GlideException
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.legue.axel.lolsummonertool.R
import com.legue.axel.lolsummonertool.database.model.champion.Spell
import com.legue.axel.lolsummonertool.database.model.champion.SpellImage
import com.legue.axel.lolsummonertool.database.viewmodel.SpellViewModel
import com.legue.axel.lolsummonertool.utils.ImageUtils
import com.legue.axel.lolsummonertool.utils.Utils
import com.legue.axel.lolsummonertool.wiki.activity.WikiChampionInformations
import kotlinx.android.synthetic.main.adapter_champion_spell.view.*
import kotlin.math.roundToInt

class ChampionSpellAdapter(private val mContext: Context, private val mSpells: List<Spell>, private val mActivity: WikiChampionInformations) : RecyclerView.Adapter<ChampionSpellAdapter.ChampionSpellHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, i: Int): ChampionSpellHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.adapter_champion_spell, parent, false)
        return ChampionSpellHolder(v)
    }

    override fun onBindViewHolder(holder: ChampionSpellHolder, i: Int) {
        val spell = mSpells[i]


        val spellViewModel = ViewModelProviders.of(mActivity).get(SpellViewModel::class.java)
        spellViewModel.getSpellImage(spell.id).observe(mActivity, Observer { spellImage: SpellImage? ->
            displayImage(spellImage?.full, holder.ivSpell, holder.pbSpell)

        })

        //TODO complete info
        holder.tvName.text = spell.name

        if (shouldConvertListToString(spell.cost!!)) {
            holder.tvCost.text = Utils.convertFloatListToString(spell.cost)
        } else {
            holder.tvCost.text = Math.round(spell.cost[0]).toString()
        }

        if (shouldConvertListToString(spell.range!!)) {
            holder.tvRange.text = Utils.convertFloatListToString(spell.range)
        } else {
            holder.tvRange.text = spell.range[0].roundToInt().toString()
        }
        holder.tvCooldown.text = Utils.convertFloatListToString(spell.cooldown!!)
        holder.tvLore.text = Html.fromHtml(spell.description, Html.FROM_HTML_MODE_COMPACT)


    }


    override fun getItemCount(): Int {
        return mSpells.size
    }

    private fun shouldConvertListToString(list: List<Float>): Boolean {
        val value = list[0]
        for (i in list.indices) {
            if (value == list[i]) {
                //do nothing
            } else {
                return true
            }
        }
        return false
    }


    private fun displayImage(url: String?, imageView: ImageView, progressBar: ProgressBar) {

        Glide.with(mContext)
                .load(ImageUtils.buildSpellIconUrl(url))
                .listener(object : RequestListener<Drawable> {
                    override fun onLoadFailed(e: GlideException?, model: Any, target: Target<Drawable>, isFirstResource: Boolean): Boolean {
                        progressBar.visibility = View.GONE
                        return false
                    }

                    override fun onResourceReady(resource: Drawable, model: Any, target: Target<Drawable>, dataSource: DataSource, isFirstResource: Boolean): Boolean {
                        progressBar.visibility = View.GONE
                        imageView.visibility = View.VISIBLE
                        return false
                    }
                })
                .error(R.drawable.ic_placeholder_black_24dp)
                .diskCacheStrategy(DiskCacheStrategy.RESOURCE)
                .into(imageView)
    }

    inner class ChampionSpellHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var ivSpell: ImageView = itemView.iv_spell
        var tvName: TextView = itemView.tv_spell_name
        var tvRange: TextView = itemView.tv_range
        var tvCost: TextView = itemView.tv_cost
        var tvCooldown: TextView = itemView.tv_cooldown
        var tvLore: TextView = itemView.tv_description
        var pbSpell: ProgressBar = itemView.pb_spell
    }

    companion object {
        private val TAG = ChampionSpellAdapter::class.java.name
    }
}
